File: apps/ios/Vostok/Core/MultiServer/ServerEntry.swift
import Foundation

/// A configured server connection, matching the web's `ServerEntry` type
/// from `apps/web/src/lib/multi-server.ts`.
struct ServerEntry: Codable, Identifiable, Equatable {
    /// Unique identifier (format: `srv_<uuid>`)
    let id: String
    /// User-facing server name
    var label: String
    /// Normalized server URL (e.g. `https://chat.example.com`)
    var url: String
    /// Hex color for server badge in chat list (e.g. `#FF5500`)
    var color: String
    /// Auth credentials (nil until authenticated)
    var auth: ServerAuth?
    /// Crypto device state (nil until registered)
    var device: ServerDevice?
    /// Server capabilities metadata
    var serverInfo: ServerInfoDTO?
    /// ISO 8601 timestamp of last successful connection
    var lastConnectedAt: String?
    /// Display order (0-based)
    var sortOrder: Int
    /// Whether this server should auto-connect
    var enabled: Bool

    /// Creates a new server entry with a generated ID.
    static func create(
        url: String,
        label: String,
        color: String = "#FF5500"
    ) -> ServerEntry {
        ServerEntry(
            id: "srv_\(UUID().uuidString.lowercased().prefix(12))",
            label: label,
            url: normalizeServerURL(url),
            color: color,
            auth: nil,
            device: nil,
            serverInfo: nil,
            lastConnectedAt: nil,
            sortOrder: 0,
            enabled: true
        )
    }
}

struct ServerAuth: Codable, Equatable {
    let token: String
    let userID: String
    let username: String
    let deviceID: String
}

struct ServerDevice: Codable, Equatable {
    let deviceID: String
    let identityPublicKey: String
}

struct ServerInfoDTO: Codable, Equatable {
    let name: String
    let version: String
    let authMode: String?

    enum CodingKeys: String, CodingKey {
        case name, version
        case authMode = "auth_mode"
    }
}

// MARK: – QualifiedChatId

/// A chat ID qualified with its server ID, format: `serverId::chatId`.
/// Matches the web's `QualifiedChatId` type from `multi-server.ts`.
typealias QualifiedChatId = String

extension QualifiedChatId {
    static func qualify(serverId: String, chatId: String) -> QualifiedChatId {
        "\(serverId)::\(chatId)"
    }
}

enum QualifiedChatIdParser {
    static func isQualified(_ value: String?) -> Bool {
        guard let value else { return false }
        return value.contains("::")
    }

    static func parse(_ qualifiedChatId: String) -> (serverId: String, chatId: String) {
        guard let separatorRange = qualifiedChatId.range(of: "::") else {
            return (serverId: "", chatId: qualifiedChatId)
        }
        let serverId = String(qualifiedChatId[qualifiedChatId.startIndex..<separatorRange.lowerBound])
        let chatId = String(qualifiedChatId[separatorRange.upperBound...])
        return (serverId: serverId, chatId: chatId)
    }

    static func serverId(from chatId: String?) -> String? {
        guard let chatId, isQualified(chatId) else { return nil }
        return parse(chatId).serverId
    }

    static func rawChatId(from chatId: String?) -> String? {
        guard let chatId else { return nil }
        if !isQualified(chatId) { return chatId }
        return parse(chatId).chatId
    }
}

// MARK: – ServerConnectionStatus

enum ServerConnectionStatus: String, Equatable {
    case disconnected
    case connecting
    case connected
    case offline
    case authRequired = "auth_required"
    case error
}

// MARK: – MergedChatSummary

/// A chat summary enriched with server metadata for unified display.
struct MergedChatSummary: Identifiable, Equatable {
    let chat: ChatDTO
    let rawId: String
    let serverId: String
    let qualifiedId: QualifiedChatId
    let serverLabel: String
    let serverColor: String
    let serverUrl: String

    var id: QualifiedChatId { qualifiedId }

    static func from(server: ServerEntry, chat: ChatDTO) -> MergedChatSummary {
        MergedChatSummary(
            chat: chat,
            rawId: chat.id,
            serverId: server.id,
            qualifiedId: .qualify(serverId: server.id, chatId: chat.id),
            serverLabel: server.label,
            serverColor: server.color,
            serverUrl: server.url
        )
    }
}

/// Sorts merged chats by latest message timestamp (newest first).
func sortMergedChats(_ chats: [MergedChatSummary]) -> [MergedChatSummary] {
    chats.sorted { left, right in
        let leftTime = left.chat.latestMessageAt ?? ""
        let rightTime = right.chat.latestMessageAt ?? ""
        return rightTime < leftTime
    }
}

// MARK: – URL normalization

func normalizeServerURL(_ url: String) -> String {
    var trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty { return "" }

    // Add https:// if no scheme
    if !trimmed.contains("://") {
        trimmed = "https://\(trimmed)"
    }

    guard var components = URLComponents(string: trimmed) else {
        return trimmed.replacingOccurrences(of: #"/+$"#, with: "", options: .regularExpression)
    }

    // Reject scheme-only URLs or those without a valid host
    guard let host = components.host, !host.isEmpty else {
        return ""
    }

    components.fragment = nil
    components.path = components.path.replacingOccurrences(of: #"/+$"#, with: "", options: .regularExpression)

    return components.string ?? trimmed
}

/// Derives a WebSocket URL from a server base URL.
func socketURL(from baseURL: String) -> URL? {
    guard var components = URLComponents(string: baseURL) else { return nil }
    components.scheme = components.scheme == "https" ? "wss" : "ws"
    components.path = "/socket/device/websocket"
    return components.url
}

File: apps/android/app/src/main/java/chat/vostok/android/core/multiserver/ServerEntry.kt
package chat.vostok.android.core.multiserver

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.net.URI
import java.util.UUID

/**
 * A configured server connection, matching the web's ServerEntry type
 * from apps/web/src/lib/multi-server.ts and iOS ServerEntry.swift.
 */
@JsonClass(generateAdapter = true)
data class ServerEntry(
    val id: String,
    var label: String,
    var url: String,
    var color: String,
    var auth: ServerAuth? = null,
    var device: ServerDevice? = null,
    @Json(name = "server_info") var serverInfo: ServerInfoDto? = null,
    @Json(name = "last_connected_at") var lastConnectedAt: String? = null,
    @Json(name = "sort_order") var sortOrder: Int = 0,
    var enabled: Boolean = true,
) {
    companion object {
        fun create(url: String, label: String, color: String = "#FF5500"): ServerEntry {
            return ServerEntry(
                id = "srv_${UUID.randomUUID().toString().take(12)}",
                label = label,
                url = normalizeServerUrl(url),
                color = color,
            )
        }
    }
}

@JsonClass(generateAdapter = true)
data class ServerAuth(
    val token: String,
    @Json(name = "user_id") val userId: String,
    val username: String,
    @Json(name = "device_id") val deviceId: String,
)

@JsonClass(generateAdapter = true)
data class ServerDevice(
    @Json(name = "device_id") val deviceId: String,
    @Json(name = "identity_public_key") val identityPublicKey: String,
)

@JsonClass(generateAdapter = true)
data class ServerInfoDto(
    val name: String,
    val version: String,
    @Json(name = "auth_mode") val authMode: String? = null,
)

// MARK: – Server Connection Status

enum class ServerConnectionStatus {
    DISCONNECTED,
    CONNECTING,
    CONNECTED,
    OFFLINE,
    AUTH_REQUIRED,
    ERROR,
}

// MARK: – QualifiedChatId

typealias QualifiedChatId = String

fun qualifyChatId(serverId: String, chatId: String): QualifiedChatId = "$serverId::$chatId"

fun isQualifiedChatId(value: String?): Boolean = value?.contains("::") == true

fun parseQualifiedChatId(qualifiedId: String): Pair<String, String> {
    val idx = qualifiedId.indexOf("::")
    if (idx < 0) return "" to qualifiedId
    return qualifiedId.substring(0, idx) to qualifiedId.substring(idx + 2)
}

fun serverIdFrom(chatId: String?): String? {
    if (chatId == null || !isQualifiedChatId(chatId)) return null
    return parseQualifiedChatId(chatId).first.takeIf { it.isNotEmpty() }
}

fun rawChatId(chatId: String?): String? {
    if (chatId == null) return null
    if (!isQualifiedChatId(chatId)) return chatId
    return parseQualifiedChatId(chatId).second
}

// MARK: – URL helpers

fun normalizeServerUrl(url: String): String {
    val trimmed = url.trim()
    if (trimmed.isEmpty()) return ""

    val withScheme = if (!trimmed.contains("://")) "https://$trimmed" else trimmed

    return try {
        val uri = URI(withScheme)
        val host = uri.host
        if (host.isNullOrEmpty()) return ""
        val path = uri.path?.trimEnd('/') ?: ""
        val port = if (uri.port > 0 && uri.port != 443 && uri.port != 80) ":${uri.port}" else ""
        "${uri.scheme}://$host$port$path"
    } catch (_: Exception) {
        trimmed.trimEnd('/')
    }
}

fun socketUrlFrom(baseUrl: String): String? {
    val normalized = normalizeServerUrl(baseUrl)
    if (normalized.isEmpty()) return null
    return try {
        val uri = URI(normalized)
        val scheme = if (uri.scheme == "https") "wss" else "ws"
        val host = uri.host ?: return null
        val port = if (uri.port > 0 && uri.port != 443 && uri.port != 80) ":${uri.port}" else ""
        "$scheme://$host$port/socket/device/websocket"
    } catch (_: Exception) {
        null
    }
}
